File: benchmarks/poly/treealloc/treealloc.go
package main

import (
	"fmt"
	"os"
	"time"
)

type Node struct {
	left, right *Node
}

func makeTree(depth int) *Node {
	if depth == 0 {
		return nil
	}
	return &Node{left: makeTree(depth - 1), right: makeTree(depth - 1)}
}

func nodeCount(n *Node) int {
	if n == nil {
		return 0
	}
	return 1 + nodeCount(n.left) + nodeCount(n.right)
}

func main() {
	t0 := time.Now()
	minDepth := 4
	maxDepth := 14

	stretch := makeTree(maxDepth + 1)
	fmt.Println(nodeCount(stretch))

	longLived := makeTree(maxDepth)

	totalCheck := 0
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << (maxDepth - depth + minDepth)
		check := 0
		for i := 0; i < iterations; i++ {
			check += nodeCount(makeTree(depth))
		}
		totalCheck += check
	}

	fmt.Println(nodeCount(longLived))
	fmt.Println(totalCheck)
	fmt.Fprintf(os.Stderr, "bench_ns=%d\n", time.Since(t0).Nanoseconds())
}

File: benchmarks/poly/treealloc/treealloc.rs
struct Node {
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn make_tree(depth: i64) -> Option<Box<Node>> {
    if depth == 0 {
        return None;
    }
    Some(Box::new(Node {
        left: make_tree(depth - 1),
        right: make_tree(depth - 1),
    }))
}

fn node_count(n: &Option<Box<Node>>) -> i64 {
    match n {
        None => 0,
        Some(node) => 1 + node_count(&node.left) + node_count(&node.right),
    }
}

fn main() {
    let t0 = std::time::Instant::now();
    let min_depth = 4;
    let max_depth = 14;

    let stretch = make_tree(max_depth + 1);
    println!("{}", node_count(&stretch));

    let long_lived = make_tree(max_depth);

    let mut total_check: i64 = 0;
    let mut depth = min_depth;
    while depth <= max_depth {
        let iterations = 1i64 << (max_depth - depth + min_depth);
        let mut check: i64 = 0;
        for _ in 0..iterations {
            check += node_count(&make_tree(depth));
        }
        total_check += check;
        depth += 2;
    }

    println!("{}", node_count(&long_lived));
    println!("{}", total_check);
    eprintln!("bench_ns={}", t0.elapsed().as_nanos());
}
